src/mono/mono/mini/aot-compiler.c

```diff
@@ -6111,6 +6114,115 @@ method_is_externally_callable (MonoAotCompile *acfg, MonoMethod *method)
 	}
 }
 
+#ifdef MONO_ARCH_AOT_SUPPORTED
+//---------------------------------------------------------------------------------------
+//
+// get_pinvoke_import:
+//
+// Returns whether or not module and entrypoint pinvoke information could be grabbed
+// from the MonoMethod. It populates module and entrypoint if they are not NULL. A
+// hash table is populated with key value pairs corresponding to the MonoMethod and
+// module entrypoint string array to serve as a fast path cache. The module and
+// entrypoint string data are owned by the hash table.
+//
+// Arguments:
+//  * acfg - the MonoAotCompiler instance
+//  * method - the MonoMethod to grab pinvoke scope and import information from
+//  ** module - the pointer to the module name string (owned by the hashtable)
+//  ** entrypoint - the pointer to the entrypoint name string (owned by the hashtable)
+//
+// Return Value:
+//  gboolean corresponding to whether or not module and entrypoint pinvoke information
+//  could be grabbed from the provided MonoMethod.
+//
+
+static gboolean
+get_pinvoke_import (MonoAotCompile *acfg, MonoMethod *method, const char **module, const char **entrypoint)
+{
+	MonoImage *image = m_class_get_image (method->klass);
+	MonoMethodPInvoke *piinfo = (MonoMethodPInvoke *) method;
+	MonoTableInfo *tables = image->tables;
+	MonoTableInfo *im = &tables [MONO_TABLE_IMPLMAP];
+	MonoTableInfo *mr = &tables [MONO_TABLE_MODULEREF];
+	guint32 im_cols [MONO_IMPLMAP_SIZE];
+	int module_idx;
+	char **scope_import;
+	guint32 scope_token;
+	char *module_ref_basename;
+	char **module_ref_basename_without_extension;
+
+	if (g_hash_table_lookup_extended (acfg->method_to_pinvoke_import, method, NULL, (gpointer *)&scope_import) && scope_import) {
+		if (module)
+			*module = scope_import[0];
+		if (entrypoint)
+			*entrypoint = scope_import[1];
+		return TRUE;
+	}
+
+	if (piinfo->implmap_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_IMPLMAP, piinfo->implmap_idx))
+		return FALSE;
+
+	mono_metadata_decode_row (im, piinfo->implmap_idx - 1, im_cols, MONO_IMPLMAP_SIZE);
+
+	module_idx = im_cols [MONO_IMPLMAP_SCOPE];
+	if (module_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_MODULEREF, module_idx))
+		return FALSE;
+
+	scope_import = (char **) g_malloc0 (2 * sizeof (char *));
+	scope_token = mono_metadata_decode_row_col (mr, im_cols [MONO_IMPLMAP_SCOPE] - 1, MONO_MODULEREF_NAME);
+	module_ref_basename = g_path_get_basename (mono_metadata_string_heap (image, scope_token));
+	module_ref_basename_without_extension = g_strsplit (module_ref_basename, ".", -1);
```

Comment:
You should look from end of string for extension, and maybe compare to known suffixes for the platform to see if its a module file extension and only drop it if its a known module suffix, like:

```
char * module_ref_basename_extension = strrchr (module_ref_basename, '.');
if (module_ref_basename_extension) {
    const char **suffixes = mono_dl_get_so_suffixes ();
    for (int i = 0; suffixes [i] && suffixes [i][0] != '\0'; i++) {
        if (!strcmp (module_ref_basename_extension, suffixes [i])) {
            *module_ref_basename_extension= '\0';
             break;
        }
    }
}

scope_import [0] = module_ref_basename;
scope_import [1] = g_strdup_printf ("%s", mono_metadata_string_heap (image, im_cols [MONO_IMPLMAP_NAME]));

g_hash_table_insert (acfg->method_to_pinvoke_import, method, scope_import);

if (module)
    *module = scope_import [0];
if (entrypoint)
    *entrypoint = scope_import [1];

return TRUE;
```

This will reduce several allocations and frees:

No need for additional allocation done by g_strsplit
No need to free additional allocation done by g_strsplit
No need to g_strdup_printf module_ref_basename into scope_import [0]
No need to free module_ref_basename since ownership will be taken by hash table.

Reply:
How should we handle other file extensions encountered that do not match these suffixes?
I noticed that there are `module_ref_basename` with values like `libSystem.IO.Compression.Native` (entrypoint `BrotliDecoderDecompress`), so I'm seeing now we can't assume there will only be one `.` in the `module_ref_basename` (which I had previously). 

So since it seems to only matter that what is passed through `DIrectPInvokes` matches what is decorated in the corresponding `DllImport` attribute in order to generate a direct call with a symbol matching that of what is linked in the final shared library, so should the module passed in the `DirectPInvoke` be the final shared library that is generated?

Reply:
Since the file will be a shared library for target platform, if it has a file suffix, it will be the one used on the platform matching what we have in mono_dl_get_so_suffixes , if nothing matches that list we leave module name as is. The example you have with libSystem.IO.Compression.Native is module without suffix, so with the proposed change that will be left as is. That example is a little bad since it still include a platform specific prefix, not sure if NativeAOT handles that so not sure if we need to do anything with that, meaning that if user would like to add direct p/invokes on that library for both Windows and Unix user would need to add both libSystem.IO.Compression.Native and System.IO.Compression.Native to its list of direct p/invoke modules, alternative have different files depending on what platform it targets.

Reply:
> So since it seems to only matter that what is passed through `DIrectPInvokes` matches what is decorated in the corresponding `DllImport` attribute in order to generate a direct call with a symbol matching that of what is linked in the final shared library, so should the module passed in the `DirectPInvoke` be the final shared library that is generated?

The modules passed to direct pinvoke is the module in the dllimport, like libSystem.IO.Compression.Native or System.IO.Compression.Native, depending on platform.

Reply:
Accidentally pushed to the wrong branch (end-to-end) instead of this one last week, so just pushed to this now, added the suggested file extension handling

Reply:
Good, I think we can revisit this later if needed, but what we currently have will make a great start.